server.js

```diff
@@ -23,7 +13,72 @@ app.use(express.json())
 
 // Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+// route provides all endpoints
+app.get('/endpoints', (req, res) => {
+  res.json({
+    response: listEndpoints(app),
+    success: true
+  })
+})
+
+// Endpoint for all books:
+// search query title and author:
+app.get('/books', (req, res) => {
+  const { title, author } = req.query
+  
+  let books = booksData.sort((a, b) => a.bookID - b.bookID)
+
+  if (title) {
+    // checks if book title is a string (because some are not!)
+    books = books.filter((book) => typeof book.title === "string")
+    // checks for title
+    books = books.filter((book) => book.title.toLowerCase().includes(title.toLowerCase()))
+  }
+
+  if (author) {
+    // checks for author
+    books = books.filter((book) => book.authors.toLowerCase().includes(author.toLowerCase()))
+  }
+
+  if (books.length === 0) {
+    // displays 404 error message if author or title search has no results
+    res.status(404).json('Not Found')
+  } else {
+    res.json(booksData)
+  }
+})
+
+// Endpoint that shows all authors in alphabetic order:
+app.get('/books/authors/:authors/', (req, res) => {
+  const listAuthors = booksData.map((item) => item.authors).sort()
```

Comment:
`const listAuthors = Array.from(new Set(booksData.map((item) => item.authors))).sort()`
This will create a list of all authors in the Books-Data without any duplicates. It's creating a Set (object with unique items). Afterwards you convert the set into an array and sort it alphabetically.
I can explain this further, if you want :)